icepack/statistics.py

```diff
@@ -0,0 +1,183 @@
+            self.val = self.rf(x.dat)
+        except firedrake.ConvergenceError:
+            self.val = np.inf
+
+
+class _ROLSolverWrapper(pyadjoint.ROLSolver):
+    def __init__(self, problem, controls, inner_product="L2"):
+        r"""Wrapper around the ROL solver class that uses the patched Objective
+        class"""
+        super(_ROLSolverWrapper, self).__init__(
+            problem, controls, inner_product=inner_product
+        )
+        self.rolobjective = _ROLObjectiveWrapper(problem.reduced_functional)
+
+
+_default_rol_options = {
+    "Step": {
+        "Type": "Trust Region",
+        "Trust Region": {
+            "Initial Radius": -1,
+            "Subproblem Solver": "Truncated CG",
+            "Radius Growing Rate": 2.5,
+            "Step Acceptance Threshold": 0.05,
+            "Radius Shrinking Threshold": 0.05,
+            "Radius Growing Threshold": 0.9,
+            "Radius Shrinking Rate (Negative rho)": 0.0625,
+            "Radius Shrinking Rate (Positive rho)": 0.25,
+            "Sufficient Decrease Parameter": 1e-4,
+            "Safeguard Size": 1e2,
+        },
+    },
+    "Status Test": {
+        "Gradient Tolerance": 1e-4,
+        "Step Tolerance": 5e-3,
+        "Iteration Limit": 50,
+    },
+    "General": {
+        "Print Verbosity": 0,
+        "Secant": {},
+    },
+}
+
+
+class MaximumProbabilityEstimator:
+    def __init__(self, problem, solver_type="rol", **kwargs):
+        from firedrake_adjoint import Control, ReducedFunctional
+
+        self._problem = problem
+        if isinstance(problem.controls, (firedrake.Function, firedrake.Constant)):
+            self._controls = problem.controls.copy(deepcopy=True)
+        else:
+            self._controls = [field.copy(deepcopy=True) for field in problem.controls]
+
+        self._state = self.problem.simulation(self.controls)
+        self._likelihood = self.problem.likelihood(self.state)
+        self._prior = self.problem.prior(self.controls)
+        J = firedrake.assemble(self._likelihood) + firedrake.assemble(self._prior)
+
+        if isinstance(self.controls, (firedrake.Function, firedrake.Constant)):
+            reduced_objective = ReducedFunctional(J, Control(self.controls))
+        else:
+            controls = [Control(field) for field in self.controls]
+            reduced_objective = ReducedFunctional(J, controls)
+
+        if isinstance(solver_type, str) and solver_type.lower() == "rol":
+            problem_wrapper = pyadjoint.MinimizationProblem(reduced_objective)
+
+            if "rol_options" in kwargs.keys():
+                options = kwargs["rol_options"]
+            else:
+                options = deepcopy(_default_rol_options)
+
+                test = options["Status Test"]
+                test["Step Tolerance"] = kwargs.get("step_tolerance", 5e-3)
+                test["Gradient Tolerance"] = kwargs.get("gradient_tolerance", 1e-4)
+                test["Iteration Limit"] = kwargs.get("max_iterations", 50)
```

Comment:
What's going on here?

Reply:
There's a bit of indirection here because I don't want to assume that we're only going to be using ROL to do the optimization. For example, Matt Knepley is talking about making TAO hook into Firedrake and pyadjoint. So what's happening here is that I picked an outer-level naming scheme that's mostly consistent with the soon-to-be deprecated code in inverse.py and translating to ROL's naming scheme, with the understanding that I'll probably need a different translation to TAO's naming scheme or MOOLA or ipopt or whatever else we want to use. It's pretty ugly, but I think it's good enough for now.

Reply:
I think a comment that explains something along these lines ought to be in the code then

Reply:
Also in this specific case I'm wondering why you're creating a dictionary `test` which you never use

Reply:
It's to directly modify the entries of a nested sub-dictionary rather than have to refer several keys deep every time.